src/recordlinker/database/mpi_service.py

```diff
@@ -539,3 +542,102 @@ def get_orphaned_persons(
+    """
+    p1 = orm.aliased(models.Patient, name="p1")
+    p2 = orm.aliased(models.Patient, name="p2")
+    random_pairs = (
+        expression.select(
+            p1.id.label("patient_1_id"),
+            p2.id.label("patient_2_id")
+        ).join_from(
+            p1,
+            p2,
+            expression.and_(
+                p1.person_id == p2.person_id,
+                p1.id < p2.id,
+                ~p1.person_id.is_(None)
+            )
+        ).order_by(
+            func.random()
+        ).limit(
+            n_pairs
+        ).cte()
+    )
+    sample = (
+        expression.select(
+            p1.person_id,
+            random_pairs.c.patient_1_id,
+            random_pairs.c.patient_2_id,
+            p1.data.label("patient_data_1"),
+            p2.data.label("patient_data_2")
+        ).join_from(
+            random_pairs, p1, p1.id == random_pairs.c.patient_1_id
+        ).join(
+            p2, p2.id == random_pairs.c.patient_2_id
+        )
+    )
+    return session.execute(sample).all()
+
+
+def generate_non_match_tuning_samples(
+        session: orm.Session,
+        sample_size: int,
+        n_pairs: int
+    ) -> typing.Sequence[typing.Tuple]:
+    """
+    Creates a sample of known "non match" pairs of patient records of 
+    size n_pairs using previously labeled data. The complete collection
+    of patient data is first randomly sub-sampled if there are more than
+    100k patient rows (since a random sample from a random sample is
+    equivalent to randomly sampling the first group), and then pairs
+    are randomly generated until the desired number of non-matches is hit.
+    """
+    # Start with a large sub-sample from which we'll draw pairs
+    query = expression.select(
+        models.Patient.id,
+        models.Patient.person_id,
+        models.Patient.data
+    ).where(
+        ~models.Patient.person_id.is_(None)
+    ).order_by(
+        func.random()
+    ).limit(
+        sample_size
+    )
+    sample = list(session.execute(query).all())
+
+    # Now, combinatorially build up the negative pairs by
+    # randomly selecting two records, making sure they don't
+    # match, and then storing them as a pair
+    already_seen = set()
+    neg_pairs: list[tuple] = []
+    while len(neg_pairs) < n_pairs:
+        idx_1 = random.randint(0, len(sample) - 1)
+        idx_2 = random.randint(0, len(sample) - 1)
+
+        # Can't make a record pair from one record
+        if idx_1 != idx_2:
+
+            # We'll use the convention of making the "smaller" ID the
+            # first index, easier for set tracking
+            record_row_1 = sample[min(idx_1, idx_2)]
+            record_row_2 = sample[max(idx_1, idx_2)]
+
+            # If the person_ids don't agree, this is a valid neg pair
+            if record_row_1[1] != record_row_2[1]:
+                if (record_row_1[0], record_row_2[0]) not in already_seen:
+                    already_seen.add((record_row_1[0], record_row_2[0]))
```

Comment:
What if, in one iteration of the loop we pull out pat_id 5 and 10.  Then a bit later we pull out pat_id 10 and 5.   I think we'd want to skip adding the second pair as we already saw that combination?

Reply:
Good catch. The min_idx lines above were intended to catch that, but that's the min index into the selection list rather than the minimum of the patient IDs, which is what I actually wanted to test. Will get that straightened out and then the already_seen set should account for that.